// @progress/telerik-angular-report-viewer/dist/cjs/telerik-reporting.module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TelerikReportingModule = void 0;
const core_1 = require("@angular/core");
const common_1 = require("@angular/common");
const telerik_report_viewer_component_1 = require("./telerik-report-viewer.component");
const i0 = require("@angular/core");
/**
 * Represents the [NgModule](https://angular.io/docs/ts/latest/guide/ngmodule.html)
 * definition for the Telerik Report Viewer component.
 *
 * @example
 *
 * ```ts-no-run
 * // Import the Grid module
 * import { TelerikReportingModule } from './telerik-reporting/telerik-reporting.module';
 *
 * import { NgModule } from '@angular/core';
 *
 * // Import the app component
 * import { AppComponent } from './app.component';
 *
 * // Define the app module
 * @@NgModule({
 *     declarations: [AppComponent], // declare app component
 *     imports:      [BrowserModule, TelerikReportingModule], // import Grid module
 *     bootstrap:    [AppComponent]
 * })
 * export class AppModule {}
 *
 * ```
 */
class TelerikReportingModule {
}
exports.TelerikReportingModule = TelerikReportingModule;
TelerikReportingModule.ɵfac = function TelerikReportingModule_Factory(t) { return new (t || TelerikReportingModule)(); };
TelerikReportingModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TelerikReportingModule });
TelerikReportingModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [common_1.CommonModule] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TelerikReportingModule, [{
        type: core_1.NgModule,
        args: [{
                imports: [common_1.CommonModule],
                exports: [telerik_report_viewer_component_1.TelerikReportViewerComponent],
                declarations: [telerik_report_viewer_component_1.TelerikReportViewerComponent],
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TelerikReportingModule, { declarations: [telerik_report_viewer_component_1.TelerikReportViewerComponent], imports: [common_1.CommonModule], exports: [telerik_report_viewer_component_1.TelerikReportViewerComponent] }); })();
