// @progress/telerik-jquery-report-viewer/dist/cjs/enums.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const ViewModes = {
  INTERACTIVE: "INTERACTIVE",
  PRINT_PREVIEW: "PRINT_PREVIEW"
};
const PrintModes = {
  AUTO_SELECT: "AUTO_SELECT",
  FORCE_PDF_PLUGIN: "FORCE_PDF_PLUGIN",
  FORCE_PDF_FILE: "FORCE_PDF_FILE"
};
const PageModes = {
  SINGLE_PAGE: "SINGLE_PAGE",
  CONTINUOUS_SCROLL: "CONTINUOUS_SCROLL"
};
const ScaleModes = {
  FIT_PAGE_WIDTH: "FIT_PAGE_WIDTH",
  FIT_PAGE: "FIT_PAGE",
  SPECIFIC: "SPECIFIC"
};
const ParameterTypes = {
  INTEGER: "System.Int64",
  FLOAT: "System.Double",
  STRING: "System.String",
  DATETIME: "System.DateTime",
  BOOLEAN: "System.Boolean"
};
const ParameterEditorTypes = {
  COMBO_BOX: "COMBO_BOX",
  LIST_VIEW: "LIST_VIEW"
};
const ParametersAreaPositions = {
  RIGHT: "RIGHT",
  LEFT: "LEFT",
  TOP: "TOP",
  BOTTOM: "BOTTOM"
};
const DocumentMapAreaPositions = {
  RIGHT: "RIGHT",
  LEFT: "LEFT"
};

exports.DocumentMapAreaPositions = DocumentMapAreaPositions;
exports.PageModes = PageModes;
exports.ParameterEditorTypes = ParameterEditorTypes;
exports.ParameterTypes = ParameterTypes;
exports.ParametersAreaPositions = ParametersAreaPositions;
exports.PrintModes = PrintModes;
exports.ScaleModes = ScaleModes;
exports.ViewModes = ViewModes;
